crypto/params.c

```diff
@@ -0,0 +1,758 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include "internal/thread_once.h"
+
+#define SET_RETURN_SIZE(p, sz) \
+    if ((p)->return_size != NULL) \
+        *(p)->return_size = (sz)
+
+const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static OSSL_PARAM ossl_param_construct(const char *key, unsigned int data_type,
+                                       void *buffer, size_t buffer_size,
+                                       size_t *return_size)
+{
+    OSSL_PARAM res;
+
+    res.key = key;
+    res.data_type = data_type;
+    res.buffer = buffer;
+    res.buffer_size = buffer_size;
+    res.return_size = return_size;
+    return res;
+}
+
+int OSSL_PARAM_get_int(const OSSL_PARAM *p, int *val)
+{
+    if (val == NULL
+        || p == NULL
+        || (p->data_type != OSSL_PARAM_INTEGER
+            && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER))
+         return 0;
+    switch (p->buffer_size) {
+    case sizeof(int32_t):
+        if (sizeof(int) >= sizeof(int32_t)) {
+            *val = (int)*(const int32_t *)p->buffer;
```

Comment:
Depends on your point of view.  The functions are *defining*, insofar that their use (should) mimic the actual properties of the value in the object that holds them.

So if we short-cut the whole passage of parameter "foo" to just an assignment of one field in the object, this is what we do:

``` C
    /*
     * in the structure for |obj|, |foo| is defined like this:
     *
     *     int foo;
     */ 
    obj->foo = (int)value;
```

However, because of the way we need to handle parameters, the assigment looks like this instead:

``` C
    OSSL_PARAM_get_int(param, &obj->foo);
```

Reply:
Not sure I follow. If the value in p->buffer > INT_MAX then that's an error and we should detect that, surely?

Reply:
In the checking mechanism that we talked about (and that I assume you mention in #8454), we can certainly try to *warn* against such uses, but just as a C compiler will not generate an error for this (unless we explicitely tell it to):

``` C
    unsigned int foo;
    int bar = foo;
```

... neither should we.

Reply:
(or should we?)

Reply:
The *simple* thing that we can decide is that `OSSL_PARAM_INTEGER` should not be used against unsigned object attributes and vice versa.  I think that would be justifiable, and that only means changing this:

``` C
    if (p->data_type == OSSL_PARAM_INTEGER
        || p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
```

to this:

``` C
    if (p->data_type == OSSL_PARAM_INTEGER) {
```

or this:

``` C
    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
```

all depending on the type of the `_get_` function

Reply:
I was following the C rules for conversions which does permit integer to unsigned in a well defined manner.  I’m comfortable enforcing signedness at the moment.  Again, changing this in the future shouldn’t break the API.

Reply:
I'm not seeing any change as a result of this discussion?

Reply:
... because no change decision has actually been made.  @paulidale's argument that he was following the C coercion rules is a bit compelling...

Reply:
If you want the C coercion rules to apply the use OPENSSL_get_uint() and pass it an int as an argument. That way it is explicit what is happening. With this it is buried away and is very non-obvious.

Reply:
Good point

Reply:
Oh but wait, hold on.  What you say can't really be done the way you think, since the `_get_` functions take a `val` *pointer*, not a value.  So `OSSL_PARAM_get_uint` will cast the value from the `OSSL_PARAM` data to `unsigned int`, which means no sign extension before getting `*val` assigned.

Meanwhile:

``` C
#include <stdio.h>

int main()
{
  char foo = -32;
  unsigned int bar = foo;

  printf("foo: %d\nbar: %u\n", foo, bar);
}
```

result:

```
: ; cc -o foo foo.c
: ; ./foo
foo: -32
bar: 4294967264
```

Reply:
Wait, no, not good point...  or I'm confused, of all the options discussed, I'm not sure what you're asking for at this point.  However, since `OSSL_PARAM_get_` functions take a pointer `val`, calling `OSSL_PARAM_get_uint` with the address of an `int` will only pretend that it's a `unsigned int`, and no coercion would take place, right?

Reply:
I was looking at it the wrong way.  Meh.  So yeah, get `_get_uint` only handle `OSSL_PARAM_UNSIGNED_INTEGER` and `_get_int` only handle `OSSL_PARAM_INTEGER`

Reply:
However, that makes `OSSL_PARAM_get_double` an oddball in this context, as it can handle the data types `OSSL_PARAM_REAL`, `OSSL_PARAM_INTEGER` and `OSSL_PARAM_UNSIGNED_INTEGER`.  Should it be limited to `OSSL_PARAM_REAL`?

Reply:
I'd say yes.

Reply:
I’d have preferred not to. Reals can support the full range of 32bit ints, so they should be allowed. 64 bit integers can exceed the real mantissa by eleven bits, but they cannot overflow a real’s range. Given reals are approximate, I didn’t see a major issue allowing these too.

Reply:
> However, since `OSSL_PARAM_get_` functions take a pointer `val`, calling `OSSL_PARAM_get_uint` with the address of an `int` will only pretend that it's a `unsigned int`, and no coercion would take place, right?

A compile time type incompatibility error would be my expectation. Adding a cast would fix the error and result in binary copy. That is the C coercion between signed and unsigned integers of the same size.

Casting a pointer to a value of a different size will go awry.